### PR TITLE
feat(singularity): store and run .sif images

### PR DESCRIPTION
**Pull request by jemten**
### This PR adds | fixes:

- Trying to avoid docker pull limit by pulling images and storing singularity images as .sif files, #1945

### How to test:

- Automatic and continuous test suite

### Expected outcome:
- [ ] Installation, unit and integration test suite pass

### Review:
- [ ] Code review
- [ ] Tests pass

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes. If applicable record manual test results in PR header
- [ ] **MINOR** - when you add functionality in a backwards compatible manner. If applicable record manual test results in PR header
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: lib/MIP/Environment/Container.pm [hunk 1]
```diff
@@ -242,6 +248,68 @@ sub get_recipe_executable_bind_path {
...
+
+    my @error_messages;
+
+  CONTAINER:
+    while ( my ( $container, $container_param_href ) = each %{$container_href} ) {
+
+        ## Only run check for .sif file
+        next CONTAINER if ( not $constraint{is_sif}->( $container_param_href->{uri} ) );
+
+        my ( $ok, $error_message ) = check_filesystem_objects_existance(
```
Comment (henrikstranneheim) [inline lib/MIP/Environment/Container.pm:288]:
```suggestion
        my ( $is_ok, $error_message ) = check_filesystem_objects_existance(
```

File: lib/MIP/Environment/Container.pm [hunk 2]
```diff
@@ -242,6 +248,68 @@ sub get_recipe_executable_bind_path {
...
+            {
+                object_name    => $container_param_href->{uri},
+                parameter_name => $container,
+                object_type    => q{executable_file},
+            }
+        );
+
+        next CONTAINER if $ok;
```
Comment (henrikstranneheim) [inline lib/MIP/Environment/Container.pm:296]:
```suggestion
        next CONTAINER if $is_ok;
```

File: lib/MIP/Environment/Container.pm [hunk 3]
```diff
@@ -242,6 +248,68 @@ sub get_recipe_executable_bind_path {
...
+    while ( my ( $container, $container_param_href ) = each %{$container_href} ) {
+
+        ## Only run check for .sif file
+        next CONTAINER if ( not $constraint{is_sif}->( $container_param_href->{uri} ) );
+
+        my ( $ok, $error_message ) = check_filesystem_objects_existance(
+            {
+                object_name    => $container_param_href->{uri},
+                parameter_name => $container,
+                object_type    => q{executable_file},
```
Comment (henrikstranneheim) [inline lib/MIP/Environment/Container.pm:292]:
```suggestion
                object_type    => q{executable_file},
                parameter_name => $container,
```
Order AB

File: t/check_installed_containers.t [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        executable => {
+            bgzip    => undef,
+            samtools => undef,
+            tabix    => undef,
+        },
+        uri => q{docker.io/clinicalgenomics/htslib:1.13},
+    },
+);
+
+## Then everything is OK
```
Comment (henrikstranneheim) [inline t/check_installed_containers.t:72]:
```suggestion
## Then sif's exists and are executable
```

File: t/mip_install.test [hunk 5]
```diff
@@ -154,6 +155,29 @@ my %process_return = child_process(
...
+## When given option to pull singularity images for local installation
+push @{$cmds_ref}, q{--singularity_local_install};
+%process_return = child_process(
+    {
+        commands_ref => $cmds_ref,
+        process_type => q{ipc_cmd_run},
+        verbose      => $verbose,
+    }
+);
+## Then return OK
```
Comment (henrikstranneheim) [inline t/mip_install.test:167]:
```suggestion
## Then return true
```

**Review by henrikstranneheim — APPROVED** [review]
Well done ⭐ !

**Comment by jemten** [comment]
The gpu version of deepvariant and deeptrio required to much disk space on hasta when the layers were squashed into a sif so I changed to the non-gpu version, since were not using the gpu:s anyway.

**Comment by jemten** [comment]
@henrikstranneheim, @ramprasadn. @projectoriented 
Just a heads up. Once this is merged to develop you will have to use the option `--install_config_file` and point it to a yaml containing paths to local singularity containers or docker adressess. The structure of this file is really similar to the one called mip_install_config.yaml. The only difference is that the new files does away with the top level key `container`.

**Review by projectoriented — COMMENTED** [review]
Viewed all! 🤩. Nice work & looking forward to use the functionality.

**Review by projectoriented — APPROVED** [review]

**Pull request merged**